### ISSUE TITLE
Expose resolver logic to plugin context

### ISSUE THREAD

**Issue by ggoodman** [evanw/esbuild#641]
When authoring plugins, I have found myself wishing I could benefit from the `esbuild` resolver logic. This logic is quite sophisticated and would be difficult to accurately reproduce and maintain. I think it would be helpful for both `onResolve` and `onLoad` plugins to be able to invoke a contextual `resolve` function that would invoke `esbuild`'s internal resolution logic. Of course, such logic would need to avoid calling back into an invoking `onResolve` hook or we risk turning the the node <--> native bus into an unintentional heat source.

**Comment by evanw** [comment]
Right now I'm thinking of eventually having something similar to Rollup's `this.resolve()` function that re-runs all of the plugin logic and then esbuild's own resolver logic. I'm hoping this will also solve for other similar things such as #501.

**Comment by SalvatorePreviti** [comment]
It would be very useful if the resolved logic is exposed and callable from the service itself. Would be amazing if it was possible to do await service.resolve(id: string, baseurl: string) [JS apis]

**Comment by ggoodman** [comment]
I feel like it would be tricky at the service level because much of the
resolving logic is based on build-specific config. If that could also be
passed in it would certainly be nifty!

**Comment by SalvatorePreviti** [comment]
One option I considered and used to quickly have a not too performant resolver is to use a load plugin to let esbuild resolve a file.
This function works and it typically takes around 1 millisecond to run. A faster implementation would be appreciated.

```js

const esbuild = require('esbuild')

/**
 * Resolves a module using esbuild module resolution
 *
 * @param {string} id
 * @returns {string} The resolved module
 */
async function esbuildResolve(id) {
  let _resolve
  const resolvedPromise = new Promise((resolve) => (_resolve = resolve))
  return Promise.race([
    resolvedPromise,
    esbuild
      .build({
        sourcemap: false,
        write: false,
        bundle: true,
        format: 'esm',
        logLevel: 'silent',
        platform: 'node',
        stdin: {
          contents: `import ${JSON.stringify(id)}`,
          loader: 'js',
          resolveDir: process.cwd(),
          sourcefile: __filename
        },
        plugins: [
          {
            name: 'resolve-main',
            setup(build) {
              build.onLoad({ filter: /.*/ }, ({ path }) => {
                id = path
                _resolve(id)
                return { contents: '' }
              })
            }
          }
        ]
      })
      .then(() => id)
  ])
}
```

**Comment by joelmoss** [comment]
This is just awesome that we can actually afford to call esbuild again to do stuff like this with only a 1ms overhead! 😃

I have a need to be able to call esbuild's internal resolution logic. So it would be better - and faster - to have this built in.

**Comment by somebee** [comment]
Are there any plans to expose the native resolver logic? I've hacked around it just like @SalvatorePreviti, but the 1ms cost for every resolve feels pretty bad. I've also tried to use a long-running builder that keeps a single build running and resolves waiting files with promises etc, but it fails with names that have already been resolved to paths, since onLoad is only ever called once per resolved file.

A simple idea I think could work would be to allow returning `{path: undefined, namespace: '... whatever ...'}` from `onResolve`, in which case esbuild would resolve the path as if it was a file, but run onLoad with the custom namespace.
